File: server/accounts/entitlements.py
"""Entitlement checks - the single server-side gate between tiers.

Nothing is gated today (one-shot purchase, every account LIFETIME); when a
freemium tier appears, premium features get `permission_classes = [...,
IsPremium]` (or a `has_premium(user)` branch) and the client's mirror gates
light up from the profile payload. The check is server-side by design: the
client copy of `plan` is presentation only - API features, sync, and pack
downloads are what actually enforce it.
"""

from __future__ import annotations

from django.utils import timezone
from rest_framework.permissions import BasePermission

from .models import Plan, UserProfile


def has_premium(user) -> bool:
    """Whether the user's plan grants full access right now. Expiry is checked
    here (not via a cron flipping rows) so a lapsed subscription degrades the
    moment it lapses, and un-lapses the moment the store webhook renews it."""
    if not user or not getattr(user, "is_authenticated", False):
        return False
    profile, _ = UserProfile.objects.get_or_create(user=user)
    if profile.plan == Plan.FREE:
        return False
    if profile.plan == Plan.PREMIUM:
        expires = profile.plan_expires_at
        return expires is None or expires > timezone.now()
    return True  # LIFETIME


class IsPremium(BasePermission):
    """DRF permission for premium-gated endpoints (unused while the app is a
    one-shot purchase - ready for the flip)."""

    message = "This feature requires an active subscription."

    def has_permission(self, request, view) -> bool:
        return has_premium(request.user)


def grant(user, plan: str, expires_at=None) -> UserProfile:
    """Set a user's plan - the hook a store webhook (RevenueCat / Play / App
    Store server notifications) will call. Kept trivial on purpose."""
    profile, _ = UserProfile.objects.get_or_create(user=user)
    profile.plan = plan
    profile.plan_expires_at = expires_at
    profile.save(update_fields=["plan", "plan_expires_at", "updated_at"])
    return profile

File: app/lib/core/entitlements.dart
/// Entitlement tiers - the client mirror of accounts/entitlements.py.
///
/// Today jibiki is a one-shot purchase: every tier resolves to full access
/// and nothing in the UI is gated. If the product pivots to freemium +
/// subscription, this file is the single flip point on the client:
///  1. set [localOnlyDefaultPlan] to 'free' (a free download can't assume
///     ownership without a store receipt),
///  2. wire the store SDK (RevenueCat / in_app_purchase) to refresh the
///     server plan, which then arrives via /auth/me and every /study/sync,
///  3. gate premium UI behind [isPremium] / the feature getters below.
///
/// The client value is presentation only - the server enforces entitlements
/// on its own checks (accounts.entitlements.IsPremium), so tampering with the
/// local flag can at most unlock buttons whose API calls will 403.
library;

import '../models/user.dart';

/// What a user without an account (local-only) is entitled to. 'lifetime'
/// while the app itself is paid up-front; flip to 'free' under freemium.
const String localOnlyDefaultPlan = 'lifetime';

class Entitlements {
  const Entitlements({required this.plan, this.expiresAt});

  /// free | premium | lifetime
  final String plan;
  final DateTime? expiresAt;

  factory Entitlements.of(UserProfile? profile, {bool localOnly = false}) {
    if (profile == null && localOnly) {
      return const Entitlements(plan: localOnlyDefaultPlan);
    }
    return Entitlements(
      plan: profile?.plan ?? 'free',
      expiresAt: profile?.planExpiresAt,
    );
  }

  bool get isPremium => switch (plan) {
        'lifetime' => true,
        'premium' => expiresAt == null || expiresAt!.isAfter(DateTime.now().toUtc()),
        _ => false,
      };

  // Feature gates - deliberately the only vocabulary the UI ever uses, so a
  // future tier split is a change HERE, not a hunt through the views. All of
  // them resolve to isPremium for now.
  bool get canDownloadPacks => isPremium;
  bool get canStudyOffline => isPremium;
  bool get canUseCommunity => isPremium;
}
